### PR TITLE
Don't always escape allocations when they're written to the stack in StoreBarrierInsertionPhase

### PR DESCRIPTION
**Pull request by saambarati**
#### b5a538c2f23dc7ecb423898320c72f8befeb4d24
<pre>
Don&apos;t always escape allocations when they&apos;re written to the stack in StoreBarrierInsertionPhase
<a href="https://bugs.webkit.org/show_bug.cgi?id=242157">https://bugs.webkit.org/show_bug.cgi?id=242157</a>

Reviewed by Yusuke Suzuki.

This is an optimization after my change in 251925@main. It turns out
that we end up writing allocations out to the stack frequently in
Speedometer2, and we were always considering this an escape. This lead
us to have some small slowdowns in a few subtests.

This patch makes it so we do a basic escape analysis on these allocations
being written to the stack. Basically, the write of the allocation to the
stack is no longer an escape. Instead, we track it for the current epoch.
Only if we read from that stack slot in the same epoch do we escape the allocation.

* Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:

Canonical link: <a href="https://commits.webkit.org/251979@main">https://commits.webkit.org/251979@main</a>
</pre>

### CODE REVIEW

File: Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp [hunk 1]
```diff
@@ -431,17 +435,50 @@ class StoreBarrierInsertionPhase : public Phase {
                 // So this analysis must mark an allocation that escapes to the heap as being part of the primordial
                 // epoch.
 
-                ClobberSet writes = writeSet(m_graph, m_node);
-                if (writes.overlaps(Heap) || writes.overlaps(Stack)) {
-                    auto escape = [] (Node* node) {
+                auto readFunc = [&] (const AbstractHeap& heap) {
+                    if (heap.kind() == Stack) {
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp]: Don't we need to have `read(World)` support?

File: Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp [hunk 2]
```diff
@@ -431,17 +435,50 @@ class StoreBarrierInsertionPhase : public Phase {
...
+                    }
+                };
+
+                bool wroteHeapOrStack = false;
+                AbstractHeap preciseStackWrite;
+                auto writeFunc = [&] (const AbstractHeap& heap) {
+                    if (heap.kind() == Stack) {
+                        wroteHeapOrStack = true;
+                        if (!heap.payload().isTop())
+                            preciseStackWrite = heap;
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp]: Can it be a Vector?
For example, `LoadVarargs` / `ForwardVarargs` are doing `write(AbstractHeap(Stack, ...))` multiple times.

File: Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp [hunk 3]
```diff
@@ -431,17 +435,50 @@ class StoreBarrierInsertionPhase : public Phase {
...
+                        Vector<AbstractHeap, 4> toRemove;
+                        for (auto [trackedHeap, node] : potentialStackEscapes) {
+                            if (trackedHeap.overlaps(heap)) {
+                                node->setEpoch(Epoch());
+                                toRemove.append(trackedHeap);
+                            }
+                        }
+
+                        for (const auto& heap : toRemove)
+                            potentialStackEscapes.remove(heap);
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp]: Let's use `potentialStackEscapes.removeIf` since it is more efficient.

**Review by Constellation — APPROVED** [review]
r=me with comments.

File: Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp (hunk 1 above)
Comment (saambarati) [reply Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp]: World will overlap with `heap.overlaps(Heap)` below. Maybe I can make the code a bit more obvious, though.

File: Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp (hunk 2 above)
Comment (saambarati) [reply Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp]: Yeah, but we're only considering this for `PutStack`/`SetLocal`. Let me add a RELEASE_ASSERT that for these nodes we only see a single stack write.

File: Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp (hunk 3 above)
Comment (saambarati) [reply Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp]: Sounds good!

File: Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp (hunk 1 above)
Comment (saambarati) [reply Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp]: sorry, I thought this comment was about some other code. You're right, let me fix this.

**Pull request merged**